src/vsr/replica.zig

```diff
@@ -6802,23 +6619,35 @@ const DVCQuorum = struct {
             assert(options.quorum_nack_prepare == options.quorum_view_change);
         }
 
-        const dvcs_total = DVCQuorum.dvcs_all(dvc_quorum).len;
-        if (dvcs_total < options.quorum_view_change) return .awaiting_quorum;
+        const dvcs_all_ = DVCQuorum.dvcs_all(dvc_quorum);
+        if (dvcs_all_.len < options.quorum_view_change) return .awaiting_quorum;
 
-        const dvcs = DVCQuorum.dvcs_canonical(dvc_quorum);
-        assert(dvcs.len > 0);
-        assert(dvcs.len <= dvcs_total);
+        const dvcs_canonical_ = DVCQuorum.dvcs_canonical(dvc_quorum);
+        assert(dvcs_canonical_.len > 0);
+        assert(dvcs_canonical_.len <= dvcs_all_.len);
 
         const op_head_max = DVCQuorum.op_max_canonical(dvc_quorum);
         const op_head_min = DVCQuorum.commit_max(dvc_quorum);
 
         // Iterate the highest definitely committed op and all maybe-uncommitted ops.
         var op = op_head_min;
         const op_head = while (op <= op_head_max) : (op += 1) {
-            // Replicas with non-canonical log-views implicitly nack all uncommitted ops.
-            var nacks: usize = dvcs_total - dvcs.len;
+            const found = for (dvcs_canonical_.constSlice()) |dvc| {
+                // This DVC is canonical, but lagging far behind.
+                if (dvc.header.op < op) continue;
+
+                const headers = message_body_as_view_headers(dvc);
+                const header_index = dvc.header.op - op;
+                assert(header_index <= headers.slice.len);
+
+                const header = &headers.slice[header_index];
+                assert(header.op == op);
+
+                if (vsr.Headers.dvc_header_type(header) == .valid) break header;
+            } else null;
```

Comment:
So here we only check that we have a header for an op, not necessary the body, right? can this create an availability issue?

That is, can we be in a situation such that:

* if we looked at DVCs from the whole cluster, we would have truncated the op
* but, looking at the DVC for the quorum we've received, we don't _nack_ this header, so it gets to be a new head
* so we send an SV with a message which should have been nacked? 

Should we also perhaps check that the `found` header is not `missing`?

Reply:
Yeah, I think it is an issue under a selective partition, when only some of the messages get through, and that just checking for "is not `missing`" is not going to help. 

Step 1, R0 is the primary, it prepares 1, 2, and replicates 2, but not 1

```
R0 1 2
R1   2
R2   2
```

Step 2, view change, R1 receives DVC from itself and R0. The new head is 2. 1 is found, and is not nacked (we need R2's for that). So R1 sends `SV 1 2`.

Step 3, R0 gets partitioned in such a way that its DVCs go through, but not repairs. So any subsequent view change where R0 participates gets stuck, because it doesnt' truncate  1, but can't repair it either. 

:thinking: it seems like we need that nasty "head can go backwards" bit if we want to make this 100% bullet-proof.

Reply:
I guess it can be less weird if `R0` crashes and corrupts `1`. Then any view change where it participates becomes stuck. Although _that_ case I think we can fix with checking for !missing.

Reply:
I agree we need to check that `found` is not `missing`. (I'm thinking the DVC quorum will keep waiting for DVCs if needed).

> Step 3, R0 gets partitioned in such a way that its DVCs go through, but not repairs. So any subsequent view change where R0 participates gets stuck, because it doesnt' truncate 1, but can't repair it either.

In step 2, R1 sent `SV 1 2` — if R1 sent an SV, that means that it finished repair, so by the end of step 2, the state is

```
R0 1 2
R1 1 2
R2   2
```

But I think your example applies if R1 didn't finish repair (and didn't send a SV), since its state would still be

```
R0 1 2
R1   2
R2   2
```

In that case, a view-change with R1/R2 would truncate 1,2.
But any view change involving R0 would require a repair of 1.

But isn't this outside our fault model though? (`(f+1)/(2f+1)` faults)
I guess we could artificially delay completion of the DVC quorum to allow all non-replicas a chance to register their nacks :thinking:

Reply:
> I guess we might actually get that out-of-the-box if we wait for !missing and found?

Right, the only issue left is if R0 is partially available — participating in DVCs, but not helping with repair.
But then again, maybe that would be ok too — eventually the view-change would cycle over to R0 as primary, and it already _has_ op 1.

Reply:
>eventually the view-change would cycle over to R0 as primary, and it already has op 1.

Right, if it's not corrupted, we'll progress, and if it _is_ corrupted, !missing would cause us to wait for full DVC set&truncate. Sounds good!